contracts/v2/MarketWithAuction.sol

```diff
@@ -0,0 +1,902 @@
+            targetListing.assetContract, 
+            targetListing.tokenOwner, 
+            _listingId, 
+            offeror, 
+            quantityWanted,
+            targetListing
+        );
+    }
+
+    /// @dev Lets an auction's creator close the auction.
+    function closeAuction(
+        uint256 _listingId
+    )
+        external
+        override
+        nonReentrant
+    {
+        Listing memory targetListing = listings[_listingId];
+        Offer memory targetBid = winningBid[_listingId];
+        
+        address closer = _msgSender();
+
+        if(targetListing.startTime > block.timestamp) {
+
+            require(
+                listings[_listingId].tokenOwner == closer,
+                "Market: caller is not the listing creator."
+            );
+
+            // Auction is considered canceled if end time has passed.
+            uint256 quantityToSend = targetListing.quantity;
+            targetListing.quantity = 0;
+            targetListing.endTime = block.timestamp;
+            listings[_listingId] = targetListing;
+
+            transferListingTokens(address(this), targetListing.tokenOwner, quantityToSend, targetListing);
+
+            emit AuctionCanceled(_listingId, targetListing.tokenOwner, targetListing);
+
+            return;
+        }
+
+        require(
+            closer == targetListing.tokenOwner || closer == targetBid.offeror,
+            "Market: must be bidder or auction creator."
+        );
+        require(
+            targetListing.listingType == ListingType.Auction,
+            "Market: listing is not an auction."
+        );
+        require(
+            targetListing.endTime < block.timestamp,
+            "Market: can only close auction after it has ended."
+        );
+
+        if(_msgSender() == targetListing.tokenOwner) {
+            /**
+             * Prevent re-entrancy by setting bid's offer amount, and listing's quantity to 0 before ERC20 transfer.
+             */
+            uint256 payoutAmount = targetBid.pricePerToken * targetBid.quantityWanted;
+
+            targetListing.quantity = 0;
+            targetListing.endTime = block.timestamp;
+            listings[_listingId] = targetListing;
+
+            targetBid.pricePerToken = 0;
+            winningBid[_listingId] = targetBid;
+
+            payout(
+                address(this), 
+                targetListing.tokenOwner,
+                targetListing.currency,
+                payoutAmount,
+                targetListing
+            );
+        } else if (_msgSender() == targetBid.offeror) {
+            /**
+             * Prevent re-entrancy by setting bid's quantity to 0 before token transfer.
+             */
+            
+            uint256 quantityToSend = targetBid.quantityWanted;
+
+            targetBid.quantityWanted = 0;
+            winningBid[_listingId] = targetBid;
+
+            transferListingTokens(address(this), targetBid.offeror, quantityToSend, targetListing);
+        }
```

Comment:
do we really care about isolating the close auction? or can we just distribute to the `offeror` and `tokenOwner` in 1 call?